Fix problem with single power spectrum

matteobachetti opened the pull request:
Resolve #663

Note for reviewers: I created a test reproducing exactly the MWE by @dhuppenkothen in #663, and a similar one for cross spectra. My first solution broke other tests, so I changed slightly the logic in `fourier.py` and `crossspectrum.py` to make everything work. 
In particular, if a segment size is almost an integer number of light curve bins (less than 1% away), instead of approximating to the lower bin I approximate to the higher one. Not more than that, because a bin with a significantly lower flux could alter the statistical properties of the data. New function in `utils.py` for this

mgullik reviewed (commented):
Hi @matteobachetti 

I hope I understood what this PR is doing. 
Isn't it better to check what is in the "last" bin that we are now considering if segment size is almost an integer number of light curve bins? 
Since the segment size could be set as the GTI length, if the flux in the last bin is zero, it probably means that the GTI is set such as that bin shouldn't be included. Do you agree?

Moreover, I didn't understand what happended when there are multiple GTIs and segment size Is None. It seems that segment size becomes gti[-1, 1] - gti[0, 0] which would return an error (I hope) since the segment size is larger than all GTIs. 

I would add tests with multiple GTIs in the case segment size is not set.

dhuppenkothen commented on stingray/crossspectrum.py:
```diff
@@ -1566,6 +1566,18 @@ def _initialize_from_any_input(
+            common_gti = cross_two_gtis(data1.gti, data2.gti)
+            data1.gti = common_gti
+            data2.gti = common_gti
+            data1 = data1.apply_gtis(inplace=False)
```
Is there a reason why you don't use `inplace=True` here, since you're storing the output in exactly the same variable?

matteobachetti replied:
Because I'm creating a copy and giving it the same name, which is different from modifying the input variable. I want to avoid side effects

dhuppenkothen commented on stingray/crossspectrum.py:
```diff
@@ -1566,6 +1566,18 @@ def _initialize_from_any_input(
+            data1 = data1.apply_gtis(inplace=False)
+            data2 = data2.apply_gtis(inplace=False)
+
+            if hasattr(data1, "counts"):
```
Why is this only checked if `data1` has an attribute `counts`? Could that not also be the case for a time series that does *not* have counts?

matteobachetti replied:
Sure, you have a different behavior in that case though. The presence of counts tells you that this is coming from binned events

matteobachetti replied:
I specified that this has to be an array attribute, so, a light curve

dhuppenkothen commented on stingray/fourier.py:
```diff
@@ -2294,13 +2309,14 @@ def avg_pds_from_events(
-    n_bin = int(segment_size / dt)
-    if fluxes is None:
-        dt = segment_size / n_bin
+    binned = fluxes is not None
```
I am confused by this: if this function generates an averaged PDS from events, then why is there even a question whether it's binned or not? I'm assuming the answer has to be no?

matteobachetti replied:
The function is called from events, but it does work with any time series. I will take the chance to rename it

matteobachetti replied:
Renamed and deprecated the old name!

dhuppenkothen reviewed (approved):
LGTM overall. A couple of minor questions, but I don't think they should stop merging.

matteobachetti commented on stingray/fourier.py:
```diff
@@ -1527,12 +1534,20 @@ def get_flux_iterable_from_segments(
             cast_kind = complex
 
-    fun = _which_segment_idx_fun(binned, dt)
+    if segment_size is None:
```
Right, I documented it

matteobachetti commented:
@mgullik , @dhuppenkothen I think I've addressed all your comments. 

Quick answers to @mgullik 's question:

> Since the segment size could be set as the GTI length, if the flux in the last bin is zero, it probably means that the GTI is set such as that bin shouldn't be included. Do you agree?
> Moreover, I didn't understand what happended when there are multiple GTIs and segment size Is None. It seems that segment size becomes gti[-1, 1] - gti[0, 0] which would return an error (I hope) since the segment size is larger than all GTIs.

This is a good point, and the code takes care of it. The GTIs always account for the bin size, so ideally they go from `lc.time[0] - dt/2` to `lc.time[-1] + dt / 2`.
